### PR TITLE
Remove MediaBrowser.Text for license violations and hackiness

### LINKED ISSUE #590: Move to upstream NLangDetect
**Problem:**
A lot of code to maintain, which is bad. Seems like I found one more 3rd party thing integrated into the source tree: NLangDetect (e.g. https://github.com/jellyfin/jellyfin/blob/4b5fbaa4cd064813155ac5ef366f6e1c597d355e/Emby.Server.Implementations/TextEncoding/NLangDetect/NLangDetectException.cs#L1). Apparently original source: https://github.com/marek-stoj/NLangDetect

**Solution:**
Migrate to that 3rd party code to avoid maintaining it ourselves. If this is impossible by just using it as a NuGet dependency - fork it and make a NuGet package.

### PR DESCRIPTION
**Pull request by cvium**
**Changes**
Removed MediaBrowser.Text since it depended on three separate libs and violated the licenses of at least two of them (Apache). The third one is MPL 1.1/GPL 2.0/LGPL 2.1.

Instead, I have added https://github.com/CharsetDetector/UTF-unknown, which is GPLv2. It is able to detect less charsets, but it's better than knowingly violating the license of 2-3 other projects (ffs).

The charset detector stuff is really just used to determine which encoding flags to give to ffmpeg in case some subtitles have the wrong encoding. The other places where this module was used only used Encoding.ASCII or Encoding.UTF8, both of which are stdlibs.

**Issues**
Fixes #590
Maybe fixes https://github.com/jellyfin/jellyfin/issues/254

### CODE REVIEW

File: MediaBrowser.MediaEncoding/Subtitles/SubtitleEncoder.cs [hunk 1]
```diff
@@ -729,7 +727,7 @@ public async Task<string> GetSubtitleFileCharacterSet(string path, string langua
         {
             var bytes = await GetBytes(path, protocol, cancellationToken).ConfigureAwait(false);
 
-            var charset = _textEncoding.GetDetectedEncodingName(bytes, bytes.Length, language, true);
+            var charset = CharsetDetector.DetectFromBytes(bytes).Detected?.EncodingName;
```
Comment (JustAMan) [inline MediaBrowser.MediaEncoding/Subtitles/SubtitleEncoder.cs]: What this does? Seems to be repeating some initialization upon subsequent calls, is it what you wanted?

File: SocketHttpListener/Net/HttpListenerRequest.cs [hunk 2]
```diff
@@ -88,7 +86,7 @@ public Encoding ContentEncoding
                         }
                     }
                 }
-                return TextEncodingExtensions.GetDefaultEncoding();
+                return Encoding.UTF8;
```
Comment (JustAMan) [inline SocketHttpListener/Net/HttpListenerRequest.cs]: I am slightly weary of this change, can we be sure that default encoding is always utf-8?

File: SocketHttpListener/Net/HttpListenerResponse.Managed.cs [hunk 3]
```diff
@@ -264,7 +260,7 @@ internal void SendHeaders(bool closing, MemoryStream ms, bool isWebSocketHandsha
                 ComputeCookies();
             }
 
-            Encoding encoding = _textEncoding.GetDefaultEncoding();
+            Encoding encoding = Encoding.UTF8;
```
Comment (JustAMan) [inline SocketHttpListener/Net/HttpListenerResponse.Managed.cs]: Same about uncertainty of utf-8 being the only default.

Comment (cvium) [reply SocketHttpListener/Net/HttpListenerResponse.Managed.cs]: https://github.com/jellyfin/jellyfin/blob/c2dd5baab7d4ff2c5cf089c394e3d7b12dd7de20/SocketHttpListener/Primitives/ITextEncoding.cs

File: MediaBrowser.MediaEncoding/Subtitles/SubtitleEncoder.cs (hunk 1 above)
Comment (cvium) [reply MediaBrowser.MediaEncoding/Subtitles/SubtitleEncoder.cs]: What do you mean? EncodingName should be the name of the code page, which it uses as a parameter to ffmpeg ` encodingParam = " -sub_charenc " + encodingParam;`

Comment (cvium) [reply MediaBrowser.MediaEncoding/Subtitles/SubtitleEncoder.cs]: Here's an example: https://dotnetfiddle.net/BmBd2B

**Comment by cvium** [comment]
NTextCat is improperly licensed I think. The perl project it is a port of is LGPL2+ licensed and NTextCat is MIT.

**Comment by joshuaboniface** [comment]
Ouch, cleanup caused a lot of conflicts for this one :( I'll try to go through the ones I understand and fix them.

**Comment by joshuaboniface** [comment]
I took a stab at it, but a few of the conflicts are a little out of my element and I'm not sure what might affect what.

File: MediaBrowser.MediaEncoding/MediaBrowser.MediaEncoding.csproj [hunk 4]
```diff
@@ -17,4 +17,15 @@
     <ProjectReference Include="..\OpenSubtitlesHandler\OpenSubtitlesHandler.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Include="System.Text.Encoding.CodePages" Version="4.5.1" />
+    <PackageReference Include="UTF.Unknown" Version="1.0.0-beta1" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Reference Include="System.Text.Encoding.CodePages, Version=4.1.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
```
Comment (Bond-009) [inline MediaBrowser.MediaEncoding/MediaBrowser.MediaEncoding.csproj]: Why is this needed?

Comment (cvium) [reply MediaBrowser.MediaEncoding/MediaBrowser.MediaEncoding.csproj]: Drone complained before. It's fine after the rebase... dunno. Maybe I was missing an import.

**Pull request merged**